pkg/provisioningv2/rke2/planner/store.go

```diff
@@ -160,28 +160,36 @@ func SecretToNode(secret *corev1.Secret) (*plan.Node, error) {
 	}
 	planData := secret.Data["plan"]
 	appliedPlanData := secret.Data["appliedPlan"]
+	appliedChecksum := string(secret.Data["applied-checksum"])
+	failedChecksum := string(secret.Data["failed-checksum"])
 	output := secret.Data["applied-output"]
 	appliedPeriodicOutput := secret.Data["applied-periodic-output"]
 	probes := secret.Data["probe-statuses"]
 	failureCount := secret.Data["failure-count"]
-	maxFailures := secret.Data["max-failures"]
+	maxFailures := secret.Data["max-failures"] // max-failures is set by Rancher, if it is not set, it will default to -1 (on the system-agent side) and attempt to apply forever
 
-	if len(failureCount) > 0 && len(maxFailures) > 0 {
+	planApplied := bytes.Equal(planData, appliedPlanData)
+
+	result.Failed = false
+
+	if len(failureCount) > 0 && planApplied && appliedChecksum == failedChecksum {
 		failureCount, err := strconv.Atoi(string(failureCount))
 		if err != nil {
 			return nil, err
 		}
-		maxFailures, err := strconv.Atoi(string(maxFailures))
-		if err != nil {
-			return nil, err
-		}
-		if failureCount >= maxFailures {
+		if failureCount > 0 {
 			result.Failed = true
-		} else {
-			result.Failed = false
+			if len(maxFailures) > 0 {
+				maxFailures, err := strconv.Atoi(string(maxFailures))
+				if err != nil {
+					return nil, err
+				}
+				if failureCount < maxFailures {
```

Comment:
Just to clarify here: if `maxFailures == -1`, then we will consider the plan to be failed even though it will retry forever?

Reply:
I added a `failure-threshold` in addition to keeping `max-failures` around.